File: instrumentation/http4s-blaze-server-2.12_0.21/src/main/scala/com/nr/instrumentation/http4s/ResponseWrapper.scala
package com.nr.instrumentation.http4s

import com.newrelic.api.agent.{ExtendedResponse, HeaderType}
import org.http4s.{Header, Response}
import org.http4s.util.CaseInsensitiveString
import java.lang.{Long => JLong}
import scala.util.Try


class ResponseWrapper[F[_]](val response: Response[F]) extends ExtendedResponse {

  def getStatus: Int =
    response.status.code

  def getStatusMessage: String =
    response.status.reason

  def getContentType: String =
    response.headers.get(CaseInsensitiveString("Content-Type"))
            .map(_.value)
            .orNull

  def getHeaderType: HeaderType =
    HeaderType.HTTP

  def setHeader(name: String, value: String): Unit = {
    response.headers.put(Header(name, value))
  }

  def getContentLength: Long =
    response.headers.get(CaseInsensitiveString("Content-Length")) match {
      case Some(header) => header.value.toLong
      case None => -1L
    }
}

object ResponseWrapper {
  def apply[F[_]](response: Response[F]): ExtendedResponse =
    new ResponseWrapper[F](response)
}

File: instrumentation/undertow-server-2.0.0/src/main/java/com/nr/agent/instrumentation/undertow/ResponseWrapper.java
package com.nr.agent.instrumentation.undertow;

import com.newrelic.api.agent.ExtendedResponse;
import com.newrelic.api.agent.HeaderType;
import io.undertow.server.HttpServerExchange;
import io.undertow.util.HeaderMap;
import io.undertow.util.HeaderValues;
import io.undertow.util.Headers;
import io.undertow.util.HttpString;

public class ResponseWrapper extends ExtendedResponse {
    private final long contentLength;
    private final HeaderMap headers;
    private final int statusCode;

    public ResponseWrapper(HttpServerExchange exchange) {
        contentLength = exchange.getResponseContentLength();
        headers = exchange.getResponseHeaders();
        statusCode = exchange.getStatusCode();
    }

    @Override
    public long getContentLength() {
        return contentLength;
    }

    @Override
    public HeaderType getHeaderType() {
        return HeaderType.HTTP;
    }

    @Override
    public void setHeader(String name, String value) {
        if (name != null && value != null) {
            headers.add(new HttpString(name), value);
        }
    }

    @Override
    public int getStatus() {
        return statusCode;
    }

    @Override
    public String getStatusMessage() {
        // Not available in the Undertow exchange object
        return null;
    }

    @Override
    public String getContentType() {
        if (headers.contains(Headers.CONTENT_TYPE)) {
            return headers.get(Headers.CONTENT_TYPE).get(0);
        }

        return null;
    }
}
